### ISSUE TITLE
abs_integrate leaks assumptions into enclosing context

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#3116]
Imported from SourceForge on 2024-07-07 14:48:13
Created by **[nbruin](https://sourceforge.net/u/nbruin/)** on 2013-03-02 05:36:01
Original: https://sourceforge.net/p/maxima/bugs/2557

---

The following shows that with abs_integrate loaded, evaluating a certain integral leaks an assumption into the global context. This causes wrong answers afterwards, because it allows `abs(x)` to be simplified to `x` erroneously.

~~~~
    Maxima 5.29.1 http://maxima.sourceforge.net
    using Lisp ECL 12.12.1
    Distributed under the GNU Public License. See the file COPYING.
    Dedicated to the memory of William Schelter.
    The function bug_report() provides bug reporting information.
    (%i1) load(abs_integrate);
    (%o1) /usr/local/sage/5.7/local/share/maxima/5.29.1/share/contrib/integratio/abs_integrate.mac
    (%i2) facts();
    (%o2) [kind(sinh, one_to_one), kind(log, one_to_one), kind(tanh, one_to_one), 
                                                         kind(log, increasing)]
    (%i3) defint(exp(-abs(x))/cosh(x),x,minf,inf);
    (%o3)                              2 log(2)
    (%i4) facts();
    (%o4) [kind(sinh, one_to_one), kind(log, one_to_one), kind(tanh, one_to_one), 
                                                  kind(log, increasing), x > 0]
    (%i5) defint(exp(-abs(x))/cosh(x),x,minf,inf);
    (%i6) defint(abs(x),x,-1,0);
                                            1
    (%o6)                                 - -
                                            2

    defint: integral is divergent.
     -- an error. To debug this try: debugmode(true);
~~~~

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-07 14:48:15
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2013-03-12 18:37:08
Original: https://sourceforge.net/p/maxima/bugs/2557/#5243

---

- **labels**: abs_integrate --> Lisp core
- Description has changed:

Diff:

~~~~

--- old
+++ new
@@ -1,5 +1,6 @@
 The following shows that with abs_integrate loaded, evaluating a certain integral leaks an assumption into the global context. This causes wrong answers afterwards, because it allows `abs(x)` to be simplified to `x` erroneously.
 
+~~~~
     Maxima 5.29.1 http://maxima.sourceforge.net
     using Lisp ECL 12.12.1
     Distributed under the GNU Public License. See the file COPYING.
@@ -23,3 +24,4 @@
 
     defint: integral is divergent.
      -- an error. To debug this try: debugmode(true);
+~~~~

~~~~

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-07 14:48:19
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2013-03-12 18:37:08
Original: https://sourceforge.net/p/maxima/bugs/2557/#b1b5

---

MEVALP1 leaks an assumption.

(%i1) foo : %e^(abs(uu)+uu)*(uu/abs(uu)+1)+%e^(abs(uu)-uu)*(uu/abs(uu)-1);
            abs(uu) + uu    uu             abs(uu) - uu    uu
(%o1)     %e             (------- + 1) + %e             (------- - 1)
                          abs(uu)                        abs(uu)
(%i2) :lisp (MEVALP1 '(($EQUAL) $FOO 0))
((($EQUAL)
  ((MPLUS SIMP)
   ((MTIMES SIMP) ((MEXPT SIMP) $%E ((MPLUS SIMP) ((MTIMES SIMP) -1 $UU) ((MABS SIMP) $UU)))
    ((MPLUS SIMP) -1 ((MTIMES SIMP) $UU ((MEXPT SIMP) ((MABS SIMP) $UU) -1))))
   ((MTIMES SIMP) ((MEXPT SIMP) $%E ((MPLUS SIMP) $UU ((MABS SIMP) $UU))) ((MPLUS SIMP) 1 ((MTIMES SIMP) $UU ((MEXPT SIMP) ((MABS SIMP) $UU) -1)))))
  0)
 (($EQUAL SIMP)
  ((MPLUS SIMP)
   ((MTIMES SIMP) ((MEXPT SIMP) $%E ((MPLUS SIMP) ((MTIMES SIMP) -1 $UU) ((MABS SIMP) $UU)))
    ((MPLUS SIMP) -1 ((MTIMES SIMP) $UU ((MEXPT SIMP) ((MABS SIMP) $UU) -1))))
   ((MTIMES SIMP) ((MEXPT SIMP) $%E ((MPLUS SIMP) $UU ((MABS SIMP) $UU))) ((MPLUS SIMP) 1 ((MTIMES SIMP) $UU ((MEXPT SIMP) ((MABS SIMP) $UU) -1)))))
  0))
(%i2) facts ();
(%o2)                       [notequal(abs(uu), 0)]


I can't figure out where the assumption is being introduced.

For the record, is(equal(foo, 0)) doesn't leak the assumption -- someone
(I can't see where) cleans it up, I guess.

This bug is tickled by "if notequal(df,0) ..." in
share/contrib/integration/partition.mac (line 89).

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-07 14:48:22
Created by **[kcrisman](https://sourceforge.net/u/kcrisman/)** on 2013-03-15 18:57:51
Original: https://sourceforge.net/p/maxima/bugs/2557/#ffae

---

Just for information, this was originally reported at <http://trac.sagemath.org/sage_trac/ticket/14209> and <https://groups.google.com/forum/?fromgroups=#!topic/sage-devel/jIjYSSCPrrY>

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-07 14:48:26
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2013-03-23 04:50:19
Original: https://sourceforge.net/p/maxima/bugs/2557/#465a

---

- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-07 14:48:29
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2013-03-23 04:50:20
Original: https://sourceforge.net/p/maxima/bugs/2557/#7323

---

Fixed by commit 4b8a98648d4: WITH-NEW-CONTEXT macro which is supposed to limit the scope of changes to the assume database, was buggy. New version of the macro doesn't leak assumptions. Commit pushed to origin/master and will appear in next Maxima release (5.30, scheduled for April 2013).
